sw/airborne/modules/dragspeed/dragspeed.c

```diff
@@ -0,0 +1,262 @@
+  dragspeed.coeff.y = DRAGSPEED_COEFF_Y;
+  dragspeed.filter = DRAGSPEED_FILTER;
+#ifdef DRAGSPEED_ZERO_X
+  dragspeed.zero.x = DRAGSPEED_ZERO_X;
+#endif
+#ifdef DRAGSPEED_ZERO_Y
+  dragspeed.zero.y = DRAGSPEED_ZERO_Y;
+#endif
+#if defined(DRAGSPEED_ZERO_X) && defined(DRAGSPEED_ZERO_Y)
+  dragspeed.zero_calibrated = TRUE;
+#else
+  dragspeed.zero_calibrated = FALSE;
+#endif
+  // Register callbacks
+  register_periodic_telemetry(DefaultPeriodic, PPRZ_MSG_ID_DRAGSPEED,
+      send_dragspeed);
+  AbiBindMsgIMU_ACCEL_INT32(DRAGSPEED_ACCEL_ID, &accel_ev, accel_cb);
+}
+
+bool dragspeed_calibrate_coeff(void)
+{
+  dragspeed.calibrate_coeff = TRUE;
+  return FALSE;
+}
+
+bool dragspeed_calibrate_zero(void)
+{
+  dragspeed.calibrate_zero = TRUE;
+  return FALSE;
+}
+
+bool dragspeed_is_calibrating(void)
+{
+  return dragspeed.calibrate_coeff || dragspeed.calibrate_zero;
+}
+
+static void accel_cb(uint8_t sender_id __attribute__((unused)), uint32_t stamp,
+    struct Int32Vect3 *accel)
+{
+  // Estimate current velocity
+  float vx = -(ACCEL_FLOAT_OF_BFP(accel->x) - dragspeed.zero.x)
+      / dragspeed.coeff.x;
+  float vy = -(ACCEL_FLOAT_OF_BFP(accel->y) - dragspeed.zero.y)
+      / dragspeed.coeff.y;
+  // Simple low-pass filter
+  dragspeed.vel.x += (1 - dragspeed.filter) * (vx - dragspeed.vel.x);
+  dragspeed.vel.y += (1 - dragspeed.filter) * (vy - dragspeed.vel.y);
+  // Send as ABI VELOCITY_ESTIMATE message
+#if DRAGSPEED_SEND_ABI_MESSAGE
+  if (!dragspeed.calibrate_coeff && !dragspeed.calibrate_zero) {
+    float vel_z = stateGetSpeedNed_f()->z;  // Workaround until VELOCITY_ESTIMATE contains flags to specify which axes are estimated.
+    AbiSendMsgVELOCITY_ESTIMATE(VEL_DRAGSPEED_ID, stamp, dragspeed.vel.x,
+        dragspeed.vel.y, vel_z, DRAGSPEED_R);
+  }
+#endif
+  // Perform calibration if required
+  calibrate_coeff(accel);
+  calibrate_zero(accel);
+}
+
+/**
+ * Calibrate drag coefficient by comparing accelerometer measurements to INS
+ * velocities.
+ * Should be performed with VEL_DRAGSPEED_ID set to ABI_DISABLE, but for safety
+ * the ABI messages are also disabled automatically when calibration is active.
+ *
+ * This routine requires the accelerometers to have been zeroed beforehand,
+ * otherwise it will return without changing the drag coefficient!
+ */
+static void calibrate_coeff(struct Int32Vect3 *accel)
+{
+  // Reset when new calibration is started
+  static bool calibrate_prev = FALSE;
+  static struct FloatVect2 coeff;
+  static int num_samples_x = 0;
+  static int num_samples_y = 0;
+  if (dragspeed.calibrate_coeff && !calibrate_prev) {
+    coeff.x = 0.0f;
+    coeff.y = 0.0f;
+    num_samples_x = 0;
+    num_samples_y = 0;
+  }
+  calibrate_prev = dragspeed.calibrate_coeff;
+  // Return when calibration is not active
+  if (!dragspeed.calibrate_coeff) {
+    return;
+  }
+  // Also return if zero calibration has not been performed yet
+  if (!dragspeed.zero_calibrated) {
+    fprintf(stderr,
```

Comment:
Will this compile if not using a linux board?

Reply:
Nope, added it to the hexa_lisaL airframe which resulted in compilation errors. Is there a define that I can use to recognize all linux boards or all boards that support printf? I think the warning will be helpful when diagnosing problems with calibration. If not, I will remove the printf.

Reply:
Unfortunately we don't have a standard for that. The proper solution is probably to provide a debug API for Paparazzi with debug levels and proper output depending of the hardware. But that is for an other PR.